addons/account/models/account_move.py

```diff
@@ -1,349 +1,1869 @@
+                    totals_map['total_amount_currency'] += diff_amount_currency
+                else:
+                    create_method = in_draft_mode and self.env['account.move.line'].new or self.env['account.move.line'].create
+                    rounding_lines = create_method({
+                        'name': self.invoice_cash_rounding_id.name,
+                        'debit': diff_balance > 0.0 and diff_balance or 0.0,
+                        'credit': diff_balance < 0.0 and -diff_balance or 0.0,
+                        'quantity': 1.0,
+                        'amount_currency': diff_amount_currency,
+                        'account_id': self.invoice_cash_rounding_id.account_id.id,
+                        'partner_id': self.partner_id.id,
+                        'move_id': self.id,
+                        'currency_id': currency and currency.id,
+                        'company_id': self.company_id.id,
+                        'company_currency_id': self.company_id.currency_id.id,
+                        'date_maturity': False,
+                        'is_rounding_line': True,
+                        'sequence': 9999,
+                    })
+            totals_map['total_balance'] += diff_balance
+            totals_map['total_amount_currency'] += diff_amount_currency
+            lines_map['rounding_lines'] = rounding_lines
+
+            if in_draft_mode:
+                rounding_lines._onchange_amount_currency()
+                rounding_lines._onchange_balance()
+        else:
+            self.line_ids -= rounding_lines
+            lines_map['rounding_lines'] = self.env['account.move.line']
+
+    @api.multi
+    def _recompute_payment_terms_lines(self, lines_map, totals_map):
+        ''' Compute the dynamic payment term lines of the journal entry.
+
+        :param lines_map:               The line_ids dispatched by type containing:
+            * base_lines:               The lines having a tax_ids set.
+            * tax_lines:                The lines having a tax_line_id set.
+            * terms_lines:              The lines generated by the payment terms of the invoice.
+            * rounding_lines:           The cash rounding lines of the invoice.
+        :param totals_map: A dictionary having two keys:
+            * total_balance:            Sum(balance) of each product / tax line.
+            * total_amount_currency:    Sum(amount_currency) of each product / tax line.
+        '''
+        self.ensure_one()
+        in_draft_mode = self != self._origin
+
+        today = fields.Date.context_today(self)
+
+        available_term_lines_index = 0
+        available_term_lines = lines_map['terms_lines'].sorted(lambda line: line.date_maturity or today)
+
+        # Compute the date to be used during the computation of the payment terms.
+        if self.invoice_payment_term_id:
+            terms_date = self.invoice_date or fields.Date.context_today(self)
```

Comment:
If there is a payment term, you shouldn't need to get a default value, do you? If you do, doesn't it mean that there is a bug elsewhere?

Reply:
From the helper of the field:

> If you use payment terms, the due date will be computed automatically at the generation